Implementation of offline phase

smkuls opened the pull request:
**Implement get random, get triple for offline phase**
* This uses AVSS to share random values, refines them and then return
them in an order.
* This is implemented as a blocking method which returns whenever a
random number becomes available. In the background we continuously AVSS
and process these values.
* This was tested by opening both a single as well a batch in an MPC
program.
* The same has been repeated for triples as well where a set of triples
are shared by each parties and then returned.

PENDING: Triple refinement before returning the triples.


**Remove duplicated code**
* Removed duplicated code in batch_reconstruction and
robust_reconstruction
* Remove unnecessary invocation of asyncio.ensure_future

smkuls committed:
Remove duplicated code

* Removed duplicated code in batch_reconstruction and
robust_reconstruction
* Remove unnecessary invocation of asyncio.ensure_future

smkuls committed:
Implement get random, get triple for offline phase

* This uses AVSS to share random values, refines them and then return
them in an order.
* This is implemented as a blocking method which returns whenever a
random number becomes available. In the background we continuously AVSS
and process these values.
* This was tested by opening both a single as well a batch in an MPC
program.
* The same has been repeated for triples as well where a set of triples
are shared by each parties and then returned.

PENDING: Triple refinement before returning the triples.

amiller commented on honeybadgermpc/offline.py:
```diff
@@ -0,0 +1,166 @@
+        send, recv = self.get_send_recv(f'{self.tag}-AVSS_VALUE_PROCESSOR')
+        pk, sks = dealer(n, t+1, seed=17)
+        self.avss_vale_processor = AvssValueProcessor(
+            pk, sks[my_id],
```
Vale -> value?

smkuls replied:
Changed!

amiller commented on honeybadgermpc/offline.py:
```diff
@@ -0,0 +1,176 @@
+        self.avss_value_processor.__exit__(*args)
+
+
+class RandomGenerator(PreProcessingBase):
```
This is particularly elegant - it exactly matches the collect inputs -> extract randomness description from the papers

smkuls committed:
Add example to generate randoms on the fly

* Add an example to demonstrate how random numbers can be generated
while an MPC program is running and keeps replenishing the number of
randomss available whenever the supply is exhausted.
* Update logging to module level logger so that the logs can be turned
off per file.

amiller reviewed (approved):
Everything looks good
